src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -20952,32 +20952,1366 @@ public class C
+
+            c.VerifyEmitDiagnostics(
+                // (14,13): warning CS8602: Dereference of a possibly null reference.
+                //             field2.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field2").WithLocation(14, 13),
+                // (16,13): warning CS8602: Dereference of a possibly null reference.
+                //             field4.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(16, 13));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_EnforcedInLocalFunction_Property_SetToMaybeNull()
+        {
+            var c = CreateNullableCompilation(["""
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string P1 { get; set; } = "";
+    public string? P2 { get; set; }
+    public string P3 { get; set; } = "";
+    public string? P4 { get; set; }
+
+    public void M()
+    {
+        init();
+
+        [MemberNotNull(nameof(P1), nameof(P2))]
+        void init()
+        {
+            P1.ToString(); // 1
+            P2.ToString(); // 2
+            P3.ToString();
+            P4.ToString(); // 3
+        }
+    }
+}
+""", MemberNotNullAttributeDefinition]);
+
+            c.VerifyEmitDiagnostics(
+                // (16,13): warning CS8602: Dereference of a possibly null reference.
+                //             P1.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "P1").WithLocation(16, 13),
+                // (17,13): warning CS8602: Dereference of a possibly null reference.
+                //             P2.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "P2").WithLocation(17, 13),
+                // (19,13): warning CS8602: Dereference of a possibly null reference.
+                //             P4.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "P4").WithLocation(19, 13));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_EnforcedInLocalFunction_CheckNotNullOnExit()
+        {
+            var c = CreateNullableCompilation(["""
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1 = "";
+    public string? field2;
+    public string field3 = "";
+    public string? field4;
+
+    public void M()
+    {
+        init();
+
+        [MemberNotNull(nameof(field1), nameof(field2))]
+        void init()
+        {
+        } // 1
+    }
+}
+""", MemberNotNullAttributeDefinition]);
+
+            // Note: we add a synthesized BoundReturnStatement in BindLocalFunctionStatement
+            //       using the block body as syntax, but we report the diagnostic on the closing brace in that case
+            c.VerifyEmitDiagnostics(
+                // (16,9): warning CS8774: Member 'field1' must have a non-null value when exiting.
```

Comment:
>'field1' [](http://example.com/codeflow?start=50&length=8)

Why report a warning for `field1`?

Reply:
That's the same behavior as with methods. If you put the `MemberNotNull` attribute, you must set the field to a not-null value by the time you exit the method. [sharplab](https://sharplab.io/#v2:CYLg1APgAgDABFAjAOgCIEsCGBzAdgewGcAXdAY0OQGF9gBTAQV0wBsBPQ9QgbgFgAoKAGYEAJjhUBAbwFw5CEUngAzdHRbBEcALxwARHr795ChIhgB+OKvXBRRk8LMq1GkboMP5TpVZsaAFiNZEwBtAFk6AFsAIzoAJwA5fGJEgFcWFgAKZii6fGUs/00ASgAaOFz8wuLREpKAXRDvALh0XHRiLJLmuRljEwBfOAB6EbhEXtHxgHdMeI7cbAkAZQAOAHYNgJA4SNiEuAByYsQjuCi0kjgAC0wANzo4TEr8XABaXAyWOHvWNKeMxudFwcDoAA9Ou1sMgpmM4HMFtDVpttrt9nF4sdaudLtc7o9nq8Pl9Mr9/oDgaCIVClrD+IMgA)